File: examples3d/platform3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Ground.
     */
    let ground_size = 10.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height, 0.0));
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
    colliders.insert_with_parent(collider, handle, &mut bodies);

    /*
     * Create the boxes
     */
    let num = 6;
    let rad = 0.2;

    let shift = rad * 2.0;
    let centerx = shift * num as f32 / 2.0;
    let centerz = shift * num as f32 / 2.0;

    for i in 0usize..num {
        for j in 0usize..num {
            for k in 0usize..num {
                let centery = if j >= num / 2 { 5.0 } else { 3.0 };
                let x = i as f32 * shift - centerx;
                let y = j as f32 * shift + centery;
                let z = k as f32 * shift - centerz;

                // Build the rigid body.
                let rigid_body = RigidBodyBuilder::dynamic().translation(Vector::new(x, y, z));
                let handle = bodies.insert(rigid_body);
                let collider = ColliderBuilder::cuboid(rad, rad, rad);
                colliders.insert_with_parent(collider, handle, &mut bodies);
            }
        }
    }

    /*
     * Setup a velocity-based kinematic rigid body.
     */
    let platform_body =
        RigidBodyBuilder::kinematic_velocity_based().translation(Vector::new(0.0, 1.5 + 0.8, 0.0));
    let velocity_based_platform_handle = bodies.insert(platform_body);
    let collider = ColliderBuilder::cuboid(rad * 10.0, rad, rad * 10.0);
    colliders.insert_with_parent(collider, velocity_based_platform_handle, &mut bodies);

    /*
     * Setup a position-based kinematic rigid body.
     */
    let platform_body = RigidBodyBuilder::kinematic_position_based().translation(Vector::new(
        0.0,
        3.0 + 1.5 + 0.8,
        0.0,
    ));
    let position_based_platform_handle = bodies.insert(platform_body);
    let collider = ColliderBuilder::cuboid(rad * 10.0, rad, rad * 10.0);
    colliders.insert_with_parent(collider, position_based_platform_handle, &mut bodies);

    /*
     * Setup a callback to control the platform.
     */
    testbed.add_callback(move |_, physics, _, run_state| {
        let velocity = Vector::new(
            0.0,
            (run_state.time * 2.0).cos(),
            run_state.time.sin() * 2.0,
        );

        // Update the velocity-based kinematic body by setting its velocity.
        if let Some(platform) = physics.bodies.get_mut(velocity_based_platform_handle) {
            platform.set_linvel(velocity, true);
            platform.set_angvel(Vector::new(0.0, 1.0, 0.0), true);
        }

        // Update the position-based kinematic body by setting its next position.
        if let Some(platform) = physics.bodies.get_mut(position_based_platform_handle) {
            let next_tra = platform.translation() + (-velocity * physics.integration_parameters.dt);
            let next_rot = platform.rotation();
            let delta_rot = Rotation::from_rotation_y(-0.5 * physics.integration_parameters.dt);
            let next_rot = delta_rot * next_rot;
            platform.set_next_kinematic_translation(next_tra);
            platform.set_next_kinematic_rotation(next_rot);
        }
    });

    /*
     * Run the simulation.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(Vec3::new(10.0, 5.0, 10.0), Vec3::ZERO);
}

File: bindings/python/rapier-testbed/rapier_testbed/examples3/platform3.py
"""Port of examples3d/platform3.rs."""
from __future__ import annotations

import math

import rapier3d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Platform"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 10.0
    ground_height = 0.1
    ground = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_h = bodies.insert(ground)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size), ground_h, bodies
    )

    num = 6
    rad = 0.2
    shift = rad * 2.0
    centerx = shift * num / 2.0
    centerz = shift * num / 2.0

    for i in range(num):
        for j in range(num):
            for k in range(num):
                centery = 5.0 if j >= num // 2 else 3.0
                x = i * shift - centerx
                y = j * shift + centery
                z = k * shift - centerz
                body = rp.RigidBody.dynamic().translation((x, y, z))
                h = bodies.insert(body)
                colliders.insert_with_parent(rp.Collider.cuboid(rad, rad, rad), h, bodies)

    # Kinematic platforms.
    vb_platform_h = bodies.insert(
        rp.RigidBody.kinematic_velocity_based().translation((0.0, 1.5 + 0.8, 0.0))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(rad * 10.0, rad, rad * 10.0), vb_platform_h, bodies
    )

    pb_platform_h = bodies.insert(
        rp.RigidBody.kinematic_position_based().translation((0.0, 3.0 + 1.5 + 0.8, 0.0))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(rad * 10.0, rad, rad * 10.0), pb_platform_h, bodies
    )

    def _cb(tb) -> None:
        t = tb._step_count * tb._integration_parameters.dt
        velocity = (0.0, math.cos(t * 2.0), math.sin(t) * 2.0)

        platform = tb.bodies.get(vb_platform_h)
        if platform is not None:
            platform.linvel = rp.Vec3(*velocity)
            platform.angvel = rp.Vec3(0.0, 1.0, 0.0)

        platform = tb.bodies.get(pb_platform_h)
        if platform is not None:
            dt = tb._integration_parameters.dt
            tra = platform.translation
            new_tra = (
                tra.x - velocity[0] * dt,
                tra.y - velocity[1] * dt,
                tra.z - velocity[2] * dt,
            )
            platform.translation = rp.Vec3(*new_tra)
            delta = rp.Rotation3.from_axis_angle((0.0, 1.0, 0.0), -0.5 * dt)
            platform.rotation = delta * platform.rotation

    testbed.add_callback(_cb)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((10.0, 5.0, 10.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
